src/braket/default_simulator/density_matrix_simulation.py

```diff
@@ -0,0 +1,242 @@
+        )
+
+    @staticmethod
+    def _apply_operations(
+        state: np.ndarray, qubit_count: int, operations: List[Union[GateOperation, KrausOperation]]
+    ) -> np.ndarray:
+        """ Applies the gate and noise operations to the density matrix.
+
+        Args:
+            state (np.ndarray): initial density matrix
+            qubit_count (int): number of qubit in the circuit
+            operations (List[Union[GateOperation, KrausOperation]]): list of GateOperation and
+                KrausOperation to be applied to the density matrix
+
+        Returns:
+            state (np.ndarray): output density matrix
+        """
+        dm_tensor = np.reshape(state, [2] * 2 * qubit_count)
+        for operation in operations:
+            matrix = get_matrix(operation)
+            targets = operation.targets
+
+            if targets:
+                if isinstance(operation, (GateOperation, Observable)):
+                    dm_tensor = DensityMatrixSimulation._apply_gate(
+                        dm_tensor, qubit_count, matrix, targets
+                    )
+                if isinstance(operation, KrausOperation):
+                    dm_tensor = DensityMatrixSimulation._apply_kraus(
+                        dm_tensor, qubit_count, matrix, targets
+                    )
+            elif targets is None:
+                if isinstance(operation, (GateOperation, Observable)):
+                    for qubit in range(qubit_count):
+                        dm_tensor = DensityMatrixSimulation._apply_gate(
+                            dm_tensor, qubit_count, matrix, (qubit,)
+                        )
+                if isinstance(operation, KrausOperation):
+                    for qubit in range(qubit_count):
+                        dm_tensor = DensityMatrixSimulation._apply_kraus(
+                            dm_tensor, qubit_count, matrix, (qubit,)
+                        )
+
+        return np.reshape(dm_tensor, (2 ** qubit_count, 2 ** qubit_count))
+
+    def retrieve_samples(self) -> List[int]:
+        """ Retrieves samples of states from the density matrix of the simulation,
+        based on the probabilities.
+
+        Returns:
+            List[int]: List of states sampled according to their probabilities
+            in the density matrix. Each integer represents the decimal encoding of the
+            corresponding computational basis state.
+        """
+        return np.random.choice(
+            self._density_matrix.shape[0], p=self.probabilities, size=self._shots
+        )
+
+    @property
+    def state(self) -> np.ndarray:
+        """
+        np.ndarray: The current state of the simulation.
+        """
+        return self.density_matrix
+
+    @property
+    def density_matrix(self) -> np.ndarray:
+        """
+        np.ndarray: The density matrix specifying the current state of the simulation.
+        """
+        return self._density_matrix
```

Comment:
Some users might find this confusing (I did). Could you please elaborate in the docstrings on how they would be different?